### DISCUSSION TITLE
открытие HTML с API

### DISCUSSION THREAD

**Discussion by tvideo** [totumonline/totum-ru-issues-and-discussions#294] [ОШБК!] [answered]
Приветствую. Подскажите пожалуйста, может кто знает, вопрос следующий:
Есть временная таблица, в которой есть кнопка.
Кнопка которая просто открывает HTML:
= : linkToDataHtml(title: "Рассчитать стоимость поездки"; html: #h_api_map)
  где ячейка #h_api_map берет HTML
    = : select(table: 'api_map'; field: 'name'; where: 'id' = 1)

Вот сам HTML, который там находится:
==============
```
<html>
<head>
<title>Расчет стоимости аренды микроавтобуса с водителем tour-bus.by</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<script async src="https://api-maps.yandex.ru/2.1/?lang=ru_RU&amp;apikey=apikey" type="text/javascript"></script>
<script type="text/javascript">
	  ymaps.ready(init);
	  function init() {
		// Стоимость за километр.
		var DELIVERY_TARIFF = 2.1,
			// Минимальная стоимость.
			MINIMUM_COST = 100,
			myMap = new ymaps.Map('map', {
			  center: [53.906882, 27.567233],
			  zoom: 10,
			  controls: []
			}
								 ),
			// Создадим панель маршрутизации.
			routePanelControl = new ymaps.control.RoutePanel({
			  options: {
				// Добавим заголовок панели.
				showHeader: true,
				title: 'Выберите точки начала и окончания'
			  }
			}
															),
			zoomControl = new ymaps.control.ZoomControl({
			  options: {
				size: 'small',
				float: 'none',
				position: {
				  bottom: 145,
				  right: 10
				}
			  }
			}
													   );
		// Пользователь сможет построить только автомобильный маршрут.
		routePanelControl.routePanel.options.set({
		  types: {
			auto: true}
		}
												);
		myMap.controls.add(routePanelControl).add(zoomControl);
		// Получим ссылку на маршрут.
		routePanelControl.routePanel.getRouteAsync().then(function (route) {
		  // Зададим максимально допустимое число маршрутов, возвращаемых мультимаршрутизатором.
		  route.model.setParams({
			results: 1}
								, true);
		  // Повесим обработчик на событие построения маршрута.
		  route.model.events.add('requestsuccess', function () {
			var activeRoute = route.getActiveRoute();
			if (activeRoute) {
			  // Получим протяженность маршрута.
			  var length = route.getActiveRoute().properties.get("distance"),
				  // Вычислим стоимость доставки.
				  price = calculate(Math.round(length.value / 1000)),
				  // Создадим макет содержимого балуна маршрута.
				  balloonContentLayout = ymaps.templateLayoutFactory.createClass(
					'<a href="https://crm.tour-bus.by/Remotes/apiyandex?Расстояние=' + length.text + '&Стоимость=' + price + '">Выбрать маршрут</a><br/>' + 
					'<span style="font-weight: bold; font-style: italic">Расстояние: ' + length.text + '</span>');
			  // Зададим этот макет для содержимого балуна.
			  route.options.set('routeBalloonContentLayout', balloonContentLayout);
			  // Откроем балун.
			  activeRoute.balloon.open();
			}
		  }
								);
		}
														 );
		// Функция, вычисляющая стоимость доставки.
		function calculate(routeLength) {
		  return Math.max(routeLength * DELIVERY_TARIFF, MINIMUM_COST);
		}
	  }
	</script>
	<style>
	  html, body, #map {
		width: 100%;
		height: 500px;
		padding: 0;
		margin: 0;
	  }
	</style>
  </head>
  <body>
	<div id="map">
	</div>
  </body>
</html>
```
==============

https://user-images.githubusercontent.com/32614108/206213990-6ea46c07-858f-4925-8253-83239891f84c.mp4

[Вот ссылка на страницу с данной кнопкой](https://crm.tour-bus.by/An/46)
Проблема следующая, кнопка открывает карту только после второго нажатия.  Пример на видео.
В чем может быть причина?
Спасибо

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Не увидели в другом репо — поэксперементируем в понедельник.

**Reply by ttm-alexeypolunin** [reply]
Я повторил один в один и яндекс честно говорит, что вы передаете кривой API-ключ (что видно из кода).

```
Uncaught ReferenceError: ymaps is not defined
```

5 строка:

```
<script async src="https://api-maps.yandex.ru/2.1/?lang=ru_RU&amp;apikey=YOUR_VALID_API_KEY_HERE" type="text/javascript"></script>
```

**Reply by erpic-ru** [reply]
@tvideo расскажите, пожалуйста, как из модалки вытащить данные в Тотум?

**Reply by ttm-alexeypolunin** [reply]
Насколько я помню он там вызывает `remotes`. С ним может быть такая проблема, что `remotes` выполняется от указанного пользователя и если надо вернуть данные во временную таблицу, то просто так не получится тк в Тотум есть защита от чтения данных чужих временных. Для этого в pro есть возможность указать динамического пользователя для `remotes`.

Это один вариант. **Еще есть linkToInput** и в него можно сунуть html и вот там есть возможность передать данные в скрытое поле со специальным классом и они будут прочитаны в переменную `input`.

**Reply by erpic-ru** [reply]
Спасибо, через linkToinput все получилось! К сожалению, проблема со срабатыванием не с 1-го раза после перезагрузки страницы остается во всех браузерах. Ключ могу прислать для тестов, если вдруг получится пофиксить

**Reply by erpic-ru** [reply]
Выяснилось, что проблема возникает в непоследовательной загрузке скриптов. Для этого строку с подключением к API пришлось поместить в html описание таблицы. В этом случае, все открывается и работает норм
`<script async src="https://api-maps.yandex.ru/2.1/?lang=ru_RU&amp;apikey=apikey" type="text/javascript"></script>`
